File: backend/app/strategy_engine/loc_netting.py
from dataclasses import dataclass
from decimal import Decimal
from typing import Literal


LocSide = Literal["buy", "sell"]
LocExecution = Literal["loc", "market_on_close"]


@dataclass(frozen=True)
class LocOrderInput:
    side: LocSide
    limit_price: Decimal | None
    quantity: int
    execution: LocExecution = "loc"


@dataclass(frozen=True)
class NettedLocOrder:
    side: LocSide
    limit_price: Decimal | None
    quantity: int
    execution: LocExecution = "loc"


def net_loc_orders(orders: list[LocOrderInput], tick_size: Decimal) -> list[NettedLocOrder]:
    if tick_size <= 0:
        raise ValueError("tick_size must be positive.")

    buy_by_price: dict[Decimal, int] = {}
    sell_by_price: dict[Decimal, int] = {}
    market_close_sell_quantity = 0
    for order in orders:
        if order.quantity <= 0:
            raise ValueError("quantity must be positive.")
        if order.execution == "market_on_close":
            if order.side != "sell":
                raise ValueError("Only sell orders can use market_on_close.")
            market_close_sell_quantity += order.quantity
            continue
        if order.limit_price is None or order.limit_price <= 0:
            raise ValueError("limit_price must be positive.")
        target = buy_by_price if order.side == "buy" else sell_by_price
        target[order.limit_price] = target.get(order.limit_price, 0) + order.quantity

    prices = sorted(set(buy_by_price) | set(sell_by_price), reverse=True)
    net_quantity = -sum(sell_by_price.values()) - market_close_sell_quantity
    result: list[NettedLocOrder] = []

    for price in prices:
        buy_quantity = buy_by_price.get(price, 0)
        if buy_quantity:
            next_net_quantity = net_quantity + buy_quantity
            _append_transition_orders(
                result,
                before=net_quantity,
                after=next_net_quantity,
                sell_price=price + tick_size,
                buy_price=price,
            )
            net_quantity = next_net_quantity

        sell_quantity = sell_by_price.get(price, 0)
        if sell_quantity:
            next_net_quantity = net_quantity + sell_quantity
            _append_transition_orders(
                result,
                before=net_quantity,
                after=next_net_quantity,
                sell_price=price,
                buy_price=price - tick_size,
            )
            net_quantity = next_net_quantity

    if net_quantity < 0:
        result.append(
            NettedLocOrder(
                "sell",
                None,
                -net_quantity,
                execution="market_on_close",
            )
        )
    return [order for order in result if order.quantity > 0]


def _append_transition_orders(
    result: list[NettedLocOrder],
    before: int,
    after: int,
    sell_price: Decimal,
    buy_price: Decimal,
) -> None:
    if after == before:
        return
    if before < 0 and after > 0:
        result.append(NettedLocOrder("buy", buy_price, min(after, after - before)))
        result.append(NettedLocOrder("sell", sell_price, min(-before, after - before)))
        return
    if before < 0:
        result.append(NettedLocOrder("sell", sell_price, min(-before, after - before)))
    if after > 0:
        result.append(NettedLocOrder("buy", buy_price, min(after, after - before)))

File: frontend/src/utils/locNetting.ts
export type LocOrderSide = "buy" | "sell";
export type LocOrderExecution = "loc" | "market_on_close";

export interface LocOrderInput {
  side: LocOrderSide;
  execution?: LocOrderExecution;
  limitPrice: number | null;
  quantity: number;
}

export interface NettedLocOrder extends LocOrderInput {
  execution: LocOrderExecution;
}

export function netLocOrders(orders: LocOrderInput[], tickSize: number): NettedLocOrder[] {
  if (tickSize <= 0) return [];

  const buyByPrice = new Map<number, number>();
  const sellByPrice = new Map<number, number>();
  const tickUnits = priceToUnits(tickSize, tickSize);
  let marketCloseSellQuantity = 0;

  for (const order of orders) {
    if (order.quantity <= 0) continue;
    if (order.execution === "market_on_close") {
      if (order.side === "sell") marketCloseSellQuantity += Math.trunc(order.quantity);
      continue;
    }
    if (order.limitPrice === null || order.limitPrice <= 0) continue;
    const priceUnits = priceToUnits(order.limitPrice, tickSize);
    const target = order.side === "buy" ? buyByPrice : sellByPrice;
    target.set(priceUnits, (target.get(priceUnits) ?? 0) + Math.trunc(order.quantity));
  }

  const prices = [...new Set([...buyByPrice.keys(), ...sellByPrice.keys()])].sort((left, right) => right - left);
  let netQuantity = -sumQuantities(sellByPrice) - marketCloseSellQuantity;
  const result: NettedLocOrder[] = [];

  for (const priceUnits of prices) {
    const buyQuantity = buyByPrice.get(priceUnits) ?? 0;
    if (buyQuantity > 0) {
      const nextNetQuantity = netQuantity + buyQuantity;
      appendTransitionOrders(result, {
        before: netQuantity,
        after: nextNetQuantity,
        sellPriceUnits: priceUnits + tickUnits,
        buyPriceUnits: priceUnits,
        tickSize,
      });
      netQuantity = nextNetQuantity;
    }

    const sellQuantity = sellByPrice.get(priceUnits) ?? 0;
    if (sellQuantity > 0) {
      const nextNetQuantity = netQuantity + sellQuantity;
      appendTransitionOrders(result, {
        before: netQuantity,
        after: nextNetQuantity,
        sellPriceUnits: priceUnits,
        buyPriceUnits: priceUnits - tickUnits,
        tickSize,
      });
      netQuantity = nextNetQuantity;
    }
  }

  if (netQuantity < 0) {
    result.push({
      side: "sell",
      execution: "market_on_close",
      limitPrice: null,
      quantity: -netQuantity,
    });
  }
  return result.filter((order) => order.quantity > 0);
}

export function hasCrossedLocOrders(orders: LocOrderInput[]): boolean {
  const buys = orders.filter(
    (order) => order.side === "buy" && order.quantity > 0 && order.limitPrice !== null,
  );
  const sells = orders.filter((order) => order.side === "sell" && order.quantity > 0);
  return buys.some((buy) =>
    sells.some(
      (sell) =>
        sell.execution === "market_on_close" ||
        (sell.limitPrice !== null && sell.limitPrice <= (buy.limitPrice as number)),
    ),
  );
}

export function tickSizeForSymbol(symbol: string | null | undefined): number {
  const compact = (symbol ?? "").trim().toUpperCase();
  if (compact.endsWith(".KS") || compact.endsWith(".KQ") || /^[0-9A-Z]{6}$/.test(compact)) return 1;
  return 0.01;
}

function appendTransitionOrders(
  result: NettedLocOrder[],
  params: {
    before: number;
    after: number;
    sellPriceUnits: number;
    buyPriceUnits: number;
    tickSize: number;
  },
) {
  const { before, after, sellPriceUnits, buyPriceUnits, tickSize } = params;
  if (after === before) return;
  if (before < 0 && after > 0) {
    result.push({ side: "buy", execution: "loc", limitPrice: unitsToPrice(buyPriceUnits, tickSize), quantity: Math.min(after, after - before) });
    result.push({ side: "sell", execution: "loc", limitPrice: unitsToPrice(sellPriceUnits, tickSize), quantity: Math.min(-before, after - before) });
    return;
  }
  if (before < 0) {
    result.push({ side: "sell", execution: "loc", limitPrice: unitsToPrice(sellPriceUnits, tickSize), quantity: Math.min(-before, after - before) });
  }
  if (after > 0) {
    result.push({ side: "buy", execution: "loc", limitPrice: unitsToPrice(buyPriceUnits, tickSize), quantity: Math.min(after, after - before) });
  }
}

function priceToUnits(price: number, tickSize: number): number {
  return Math.round(price / tickSize);
}

function unitsToPrice(units: number, tickSize: number): number {
  return Number((units * tickSize).toFixed(tickSize >= 1 ? 0 : 2));
}

function sumQuantities(map: Map<number, number>): number {
  return [...map.values()].reduce((sum, quantity) => sum + quantity, 0);
}
